File: rest-service/src/main/java/com/example/restservice/service/NewsService.java
package com.example.restservice.service;

import com.example.restservice.model.News;
import com.example.restservice.model.User;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;

import java.util.List;

/**
 * Interfața care definește operațiile disponibile pentru gestionarea știrilor în sistem.
 * Aceste operații includ obținerea tuturor știrilor, obținerea unei știri după ID,
 * crearea, actualizarea și ștergerea știrilor, precum și obținerea știrilor asociate unei anumite categorii
 * și a utilizatorilor care au abonamente pentru o anumită știre.
 */
public interface NewsService {
    /**
     * Obține o listă cu toate știrile din sistem.
     *
     * @return Lista cu știri
     */
    List<News> getAllNews();

    /**
     * Obține o știre după ID-ul specificat.
     *
     * @param id ID-ul știrii căutate
     * @return Știrea cu ID-ul specificat sau null dacă nu există
     */
    News getNewsById(Long id);

    /**
     * Creează o știre nouă în sistem.
     *
     * @param news Știrea care urmează să fie creată
     * @return Știrea creată
     */
    News createNews(News news);

    /**
     * Actualizează o știre existentă în sistem.
     *
     * @param id         ID-ul știrii care urmează să fie actualizată
     * @param updateNews Știrea actualizată
     * @return Știrea actualizată sau null dacă nu există o știre cu ID-ul specificat
     */
    News updateNews(Long id, News updateNews);

    /**
     * Șterge o știre existentă din sistem.
     *
     * @param id ID-ul știrii care urmează să fie ștearsă
     */
    void deleteNews(Long id);

    /**
     * Obține o listă cu știrile asociate unei anumite categorii.
     *
     * @param categoryName Numele categoriei pentru care se caută știrile
     * @return Lista cu știrile asociate categoriei specificate
     */
    List<News> getNewsByCategoryName(String categoryName);

    /**
     * Obține o listă cu utilizatorii care au abonamente pentru o anumită știre.
     *
     * @param newsId ID-ul știrii pentru care se caută utilizatorii cu abonamente
     * @return Lista cu utilizatorii care au abonamente pentru știrea specificată
     */
    List<User> getUsersByNewsId(Long newsId);

    List<News> findTop3ByOrderByPublishedAtDesc(Pageable pageable);
}

File: front-site/src/app/service/news.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Category } from './category.service';

// Definirea interfeței News
export interface News {
  newsId: number;
  category: Category;
  title: string;
  content: string;
  publishedAt: string;
  image: string;
  // Alte câmpuri relevante
}

@Injectable({
  providedIn: 'root'
})
export class NewsService {
  private baseUrl = 'http://localhost:8080/news';

  constructor(private http: HttpClient) { }

  // Metodă pentru obținerea tuturor știrilor
  getAllNews(): Observable<News[]> {
    return this.http.get<News[]>(this.baseUrl);
  }

  // Metodă pentru obținerea unei știri după ID
  getNewsById(id: number): Observable<News> {
    return this.http.get<News>(`${this.baseUrl}/${id}`);
  }

  // Metodă pentru crearea unei știri noi
  createNews(news: News): Observable<News> {
    return this.http.post<News>(this.baseUrl, news);
  }

  // Metodă pentru actualizarea unei știri existente
  updateNews(id: number, news: News): Observable<News> {
    return this.http.put<News>(`${this.baseUrl}/${id}`, news);
  }

  // Metodă pentru ștergerea unei știri după ID
  deleteNews(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  // Metodă pentru obținerea știrilor după numele categoriei
  getNewsByCategoryName(categoryName: string): Observable<News[]> {
    return this.http.get<News[]>(`${this.baseUrl}/getNewsByCategoryName`, {
      params: { categoryName }
    });
  }

  // Metodă pentru obținerea utilizatorilor după ID-ul știrii
  getUsersByNewsId(id: number): Observable<any[]> { // Definește un tip adecvat pentru User
    return this.http.get<any[]>(`${this.baseUrl}/getUsersByNewsId/${id}`);
  }
  getLatestNews(): Observable<News[]> {
    return this.http.get<News[]>(`${this.baseUrl}/latest-news`);
  }
}
